File: aoc-2025/cpp/src/day02.cpp
#include <cstddef>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

const std::string DEFAULT_INPUT = "../input/day02.txt";
const std::string DEFAULT_OUTPUT = "output/day02.txt";

using ll = long long;
using Range = std::pair<ll, ll>;

std::vector<Range> getRanges(const std::string &filename) {
  std::ifstream file(filename);
  if (!file.is_open()) {
    std::cerr << "Error opening file: " << filename << std::endl;
    return {};
  }

  std::string content, line;
  while (file >> line)
    content += line;
  std::vector<Range> ranges;
  std::stringstream ss(content);
  std::string segment;
  while (std::getline(ss, segment, ',')) {
    if (segment.empty())
      continue;
    size_t sep = segment.find("-");
    if (sep != std::string::npos) {
      ll s = std::stoll(segment.substr(0, sep));
      ll e = std::stoll(segment.substr(sep + 1));
      ranges.push_back({s, e});
    }
  }
  return ranges;
}

bool isValidPt1(ll n) {
  std::string s = std::to_string(n);
  ll len = s.length();
  if (len % 2 != 0)
    return false;
  ll mid = len / 2;
  for (ll i = 0; i < mid; i++) {
    if (s[i] != s[mid + i])
      return false;
  }
  return true;
}

bool isValidPt2(ll n) {
  std::string s = std::to_string(n);
  ll len = s.length();
  ll mid = len / 2;
  for (ll l = 1; l <= mid; l++) {
    if (len % l == 0) {
      bool match = true;
      for (ll i = l; i < len; i++) {
        if (s[i] != s[i - l]) {
          match = false;
          break;
        }
      }
      if (match)
        return true;
    }
  }
  return false;
}

ll part_one(const std::string &filename) {
  std::vector<Range> ranges = getRanges(filename);
  ll ans = 0;
  for (auto &[s, e] : ranges)
    for (ll i = s; i <= e; i++)
      if (isValidPt1(i))
        ans += i;
  return ans;
}

ll part_two(const std::string &filename) {
  std::vector<Range> ranges = getRanges(filename);
  ll ans = 0;
  for (auto &[s, e] : ranges)
    for (ll i = s; i <= e; i++)
      if (isValidPt2(i))
        ans += i;
  return ans;
}

int main(int argc, char *argv[]) {
  std::string input_file = (argc > 1) ? argv[1] : DEFAULT_INPUT;
  std::cout << "Reading from: " << input_file << std::endl;

  ll p1 = part_one(input_file);
  ll p2 = part_two(input_file);

  std::cout << "Part 1: " << p1 << std::endl;
  std::cout << "Part 2: " << p2 << std::endl;

  std::ofstream oFile(DEFAULT_OUTPUT);
  if (oFile.is_open())
    oFile << p1 << "\n" << p2 << "\n";

  return 0;
}

File: aoc-2025/rust/src/bin/day02.rs
//! https://adventofcode.com/2025/day/2
use std::env;
use std::fs;
use std::path::Path;

const DEFAULT_INPUT: &str = "";
const DEFAULT_OUTPUT: &str = "";

fn parse_ranges(content: &str) -> Vec<(i64, i64)> {
    let clean_content: String = content.chars().filter(|c| !c.is_whitespace()).collect();

    let mut ranges = Vec::new();

    for segment in clean_content.split(',') {
        if segment.is_empty() {
            continue;
        }
        if let Some((s_str, e_str)) = segment.split_once('-') {
            if let (Ok(s), Ok(e)) = (s_str.parse::<i64>(), e_str.parse::<i64>()) {
                ranges.push((s, e));
            }
        }
    }
    ranges
}

fn is_valid_pt1(n: i64) -> bool {
    let s = n.to_string();
    let len = s.len();
    if len % 2 != 0 {
        return false;
    }
    let mid = len / 2;
    s[..mid] == s[mid..]
}

fn is_valid_pt2(n: i64) -> bool {
    let s = n.to_string();
    let bytes = s.as_bytes();
    let len = bytes.len();

    for l in 1..=len / 2 {
        if len % l == 0 {
            let mut matches = true;
            for i in l..len {
                if bytes[i] != bytes[i - l] {
                    matches = false;
                    break;
                }
            }
            if matches {
                return true;
            }
        }
    }
    false
}

fn part_one(content: &str) -> i64 {
    let ranges = parse_ranges(content);
    let mut ans: i64 = 0;

    for (start, end) in ranges {
        for i in start..=end {
            if is_valid_pt1(i) {
                ans += i;
            }
        }
    }
    ans
}

fn part_two(content: &str) -> i64 {
    let ranges = parse_ranges(content);
    let mut ans: i64 = 0;

    for (start, end) in ranges {
        for i in start..=end {
            if is_valid_pt2(i) {
                ans += i;
            }
        }
    }
    ans
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let input_path = args.get(1).map(|s| s.as_str()).unwrap_or(DEFAULT_INPUT);
    println!("Reading from: {}", input_path);

    let content = fs::read_to_string(input_path).expect("Could not read input file");

    let p1 = part_one(&content);
    let p2 = part_two(&content);

    println!("Part 1: {}", p1);
    println!("Part 2: {}", p2);

    if let Some(parent) = Path::new(DEFAULT_OUTPUT).parent() {
        let _ = fs::create_dir_all(parent);
    }
    let _ = fs::write(DEFAULT_OUTPUT, format!("{}\n{}\n", p1, p2));
}
